models/intermediate/int_transform_contract_code.sql

```diff
@@ -0,0 +1,175 @@
+
+Description:
+------------
+This intermediate model combines incremental loading and transformations for contract code data.
+It processes new or updated data, calculates new fields, and performs deduplication.
+
+Load Type: Truncate and Reload
+
+Features:
+---------
+1. Uses execution_date from Airflow to determine the processing window.
+2. Fetches data till the previous day from the Staging layer, one key hash per day (latest).
+3. Calculates `contract_create_ts` and `contract_delete_ts` from staging data.
+4. Joins the calculated timestamps with the incremental data and applies transformations.
+5. Calculates a row hash using SHA256 on all relevant fields.
+6. Performs deduplication based on the row hash of the fully transformed data. This step primarily detects duplicates in consecutive rows.
+
+Usage:
+------
+Compile:
+    dbt compile --models int_transform_contract_code
+
+Run:
+    dbt run --models int_transform_contract_code --full-refresh
+    dbt run --models int_transform_contract_code
+    dbt run --models int_transform_contract_code --vars '{"execution_date": "2024-10-01T00:00:00+00:00"}'
+*/
+
+-- Set the execution date (use Airflow value, fallback to dbt default if absent)
+{% set execution_date = var('execution_date', dbt_airflow_macros.ts()) %}
+
+with
+    source_data as (
+        select
+            contract_code_hash
+            , ledger_key_hash
+            , last_modified_ledger
+            , ledger_entry_change
+            , ledger_sequence
+            , deleted
+            , closed_at
+            , n_instructions
+            , n_functions
+            , n_globals
+            , n_table_entries
+            , n_types
+            , n_data_segments
+            , n_elem_segments
+            , n_imports
+            , n_exports
+            , n_data_segment_bytes
+            , batch_id
+            , batch_run_date
+            , cast('{{ execution_date }}' as timestamp) as airflow_start_ts
+            -- Rank records to identify the latest entry
+            , row_number() over (
+                partition by ledger_key_hash, cast(closed_at as date)
+                order by closed_at desc, ledger_sequence desc
+            ) as row_num
+        from {{ ref('stg_contract_code') }}
+        where
+            -- Process only completed days up to execution_date
+            date(closed_at) < date('{{ execution_date }}')
+    )
+
+    , filtered_data as (
+        select *
+        from source_data
+        where row_num = 1  -- Only consider the latest record
+    )
+
+    , derived_data as (
+        select
+            ledger_key_hash
+            , min(case when ledger_entry_change = 0 then closed_at end) as contract_create_ts  -- Calculate contract created timestamp
+            -- Calculate contract deletion timestamp
+            , max(case when ledger_entry_change = 2 and deleted = true then closed_at end) as contract_delete_ts
+        from source_data
+        group by ledger_key_hash
```

Comment:
I think the derived data can be moved to it's own incremental table instead of regenerating the whole table every run

Reply:
- Thought about this too but since there are only couple fields that needed tracking, kept it in here. 

- This CTE is also the very reason why I had to combine an Incremental Pull Model and Transformation Model into one.

- We can definitely discuss the pros/cons of maintaining a table of its own

Reply:
Depends on how many models we want to build for these i guess.
If we create another Model/table for maintaining this information, there will 4 models 

- Incremental Model to bring in Staging data
- Incremental Model to calculate the contract created/delete ts
- Full truncate and Reload Model to perform Transformations
- SCD2 Model

vs (current design)
- Truncate and Reload Model that does take care of things Transformation Model (step 3 above)
- SCD2 Model